### PR TITLE
Add persistent disk encryption

### PR DESCRIPTION
**Pull request by frelon**
Add flags to encrypt persistent partition on install:

* encrypt-persistent: flag to enable luks encryption on persistent partition.
* enroll-passphrase: string to enroll as passphrase to unlock partition.
* enroll-key-file: key-file to enroll as key to unlock partition.

During install this will invoke cryptsetup to create the LUKS partition and during mount we use systemd-cryptsetup to attach the partition before mounting the contained filesystem.

This also introduces some changes in the grub configuration, the encrypted_volumes variable can be set in grub_oem_env during install to configure which volumes are actually encrypted.

This change is tested in the new `test-encryption` target that runs an `elemental install` with a random passphrase and tests that the persistent partition and any extra-partitions can be unlocked when booted from the recovery system.

In the future the hope is to extend this to include state and oem partition, but that will need more changes to the bootloader configuration (most notably using the `cryptomount` command)

Part of #1782

### CODE REVIEW

**Comment by jmpolom** [comment]
Are you considering supporting binding the LUKS volume to the TPM (or FIDO2 device) via systemd-cryptenroll with this? I would recommend either of these options in addition to password and key files.

**Comment by frelon** [comment]
> Are you considering supporting binding the LUKS volume to the TPM (or FIDO2 device) via systemd-cryptenroll with this? I would recommend either of these options in addition to password and key files.

Yep, this will be low hanging fruit for a follow up pr! ;)

**Comment by frelon** [comment]
I have been testing this by booting an installer with `make prepare-installer-test` and then running `elemental install /dev/vda --debug --encrypt-persistent --enroll-passphrase <passphrase>` and then rebooting.

It should also work to add a install yaml like the following:

```
install:
  target: /dev/vda
  encrypt-persistent: true
  enroll-passphrase:
    - "passphrase"
```

This config expands to:

```
install:
  target: /dev/vda
  partitions:
    persistent:
      encryption:
        name: /dev/mapper/cr_persistent
        key_slots:
        -  slot: 0
           passphrase: "passphrase"
```

which should also work, but is a bit cumbersome.

File: Dockerfile [hunk 1]
```diff
@@ -57,7 +57,8 @@ RUN ARCH=$(uname -m); \
         gptfdisk \
         patterns-microos-selinux \
         btrfsprogs \
-	snapper \
+	    snapper \
```
Comment (davidcassany) [inline Dockerfile:60]: nit: there is something odd with spaces, I guess a mixture of tabs and spaces.

File: cmd/install.go [hunk 2]
```diff
@@ -123,6 +123,7 @@ func NewInstallCmd(root *cobra.Command, addCheckRoot bool) *cobra.Command {
 	addSharedInstallUpgradeFlags(c)
 	addLocalImageFlag(c)
 	addPlatformFlags(c)
+	addEncryptionFlags(c)
```
Comment (davidcassany) [inline cmd/install.go:126]: I think those would also make sense on reset command.

File: pkg/features/embedded/grub-default-bootargs/etc/elemental/bootargs.cfg [hunk 3]
```diff
@@ -1,9 +1,12 @@
 # bootargs.cfg inherits from grub.cfg several context variables:
 #   'img' => defines the image path to boot from. Active img is statically defined, does not require a value
+#   'mode' => active/passive/recovery, mode to boot
 #   'state_label' => label of the state partition filesystem
 #   'oem_label' => label of the oem partition filesystem
 #   'recovery_label' => label of the recovery partition filesystem
 #   'snapshotter' => snapshotter type, assumes loopdevice type if undefined
+#   'snap_arg' => kernel args for snapshotter
```
Comment (davidcassany) [inline pkg/features/embedded/grub-default-bootargs/etc/elemental/bootargs.cfg:8]: nit: this is not inherited, this is not part of an _API_ this is just a local variable. I'd not list it here.

File: pkg/features/embedded/grub-default-bootargs/etc/elemental/bootargs.cfg [hunk 4]
```diff
@@ -1,9 +1,12 @@
 # bootargs.cfg inherits from grub.cfg several context variables:
 #   'img' => defines the image path to boot from. Active img is statically defined, does not require a value
+#   'mode' => active/passive/recovery, mode to boot
```
Comment (davidcassany) [inline pkg/features/embedded/grub-default-bootargs/etc/elemental/bootargs.cfg:3]: Good catch!

File: pkg/types/config.go [hunk 5]
```diff
@@ -261,6 +264,36 @@ func (i *InstallSpec) Sanitize() error {
 		}
 	}
 
+	// Setup disk encryption
+	if i.EncryptPersistent {
```
Comment (davidcassany) [inline pkg/types/config.go:268]: Shouldn´t error out if EncryptPersistent is set to true and no secret is provided?

**Review by davidcassany — COMMENTED** [review]
Looks good to me. This is cool!

I miss unit tests and including the encryption configuration in state file. So we can reset and get the persistent encrypted again as it was configured at a time or eventually re-encrypt with different parameters.

**Comment by frelon** [comment]
> Looks good to me. This is cool!
> 
> I miss unit tests and including the encryption configuration in state file. So we can reset and get the persistent encrypted again as it was configured at a time or eventually re-encrypt with different parameters.

Good point, I will have an extra look at the reset functionality with this in mind!

File: pkg/types/config.go (hunk 5 above)
Comment (frelon) [reply pkg/types/config.go:268]: > Shouldn´t error out if EncryptPersistent is set to true and no secret is provided?

I actually didn't test this edge-case.. I would expect an empty keyslot to allow anything to unlock it, but I will give it a try!

File: pkg/features/embedded/grub-default-bootargs/etc/elemental/bootargs.cfg (hunk 3 above)
Comment (frelon) [reply pkg/features/embedded/grub-default-bootargs/etc/elemental/bootargs.cfg]: Yep, removed in latest commit!

File: Dockerfile (hunk 1 above)
Comment (frelon) [reply Dockerfile]: Yep, something weird happened when resolving a merge conflict.. fixed!

**Comment by jmpolom** [comment]
Is LUKS encrypted persistent storage volumes now out of scope/off the development plan for elemental toolkit?
